beacon_node/beacon_chain/src/attestation_verification.rs

```diff
@@ -0,0 +1,890 @@
+        // aggregation bit set.
+        let num_aggreagtion_bits = attestation.aggregation_bits.num_set_bits();
+        if num_aggreagtion_bits != 1 {
+            return Err(Error::NotExactlyOneAggregationBitSet(num_aggreagtion_bits));
+        }
+
+        // Attestations must be for a known block. If the block is unknown, we simply drop the
+        // attestation and do not delay consideration for later.
+        verify_head_block_is_known(chain, &attestation)?;
+
+        let indexed_attestation = obtain_indexed_attestation(chain, &attestation)?;
+
+        let validator_index = *indexed_attestation
+            .attesting_indices
+            .first()
+            .ok_or_else(|| Error::NotExactlyOneAggregationBitSet(0))?;
+
+        /*
+         * The attestation is the first valid attestation received for the participating validator
+         * for the slot, attestation.data.slot.
+         */
+        if chain
+            .observed_attesters
+            .validator_has_been_observed(&attestation, validator_index as usize)
+            .map_err(|e| BeaconChainError::from(e))?
+        {
+            return Err(Error::PriorAttestationKnown {
+                validator_index,
+                epoch: attestation.data.target.epoch,
+            });
+        }
+
+        // The signature of attestation is valid.
+        verify_attestation_signature(chain, &indexed_attestation)?;
+
+        // Now that the attestation has been full verified, note that we have received a valid
+        // attestation from this validator.
+        //
+        // It's important to double check that the attestation still hasn't been observed, since
+        // there can be a race-condition if we receive two attestations at the same time and
+        // process them in different threads.
+        if chain
+            .observed_attesters
+            .observe_validator(&attestation, validator_index as usize)
+            .map_err(|e| BeaconChainError::from(e))?
+        {
+            return Err(Error::PriorAttestationKnown {
+                validator_index,
+                epoch: attestation.data.target.epoch,
+            });
+        }
+
+        Ok(Self {
+            attestation,
+            indexed_attestation,
+        })
+    }
+
+    /// A helper function to add this attestation to `beacon_chain.op_pool`.
+    pub fn add_to_pool(self, chain: &BeaconChain<T>) -> Result<Self, Error> {
+        chain.add_to_naive_aggregation_pool(self)
+    }
+
+    /// Returns the wrapped `attestation`.
+    pub fn attestation(&self) -> &Attestation<T::EthSpec> {
+        &self.attestation
+    }
+
+    /// Returns a mutable reference to the underlying attestation.
+    ///
+    /// Marked as `unsafe` since modifying the `IndexedAttestation` can cause the attestation to
+    /// no-longer be valid. Only for use in testing.
+    pub unsafe fn indexed_attestation_mut(&mut self) -> &mut IndexedAttestation<T::EthSpec> {
```

Comment:
I'd prefer not to mark this `unsafe`, because as you say it doesn't involve memory unsafety or undefined behaviour at all.

Seeing as this function is only used in the tests for this crate, `#[cfg(test)]` will work, and is a lot cleaner